### PR TITLE
Validate builtin attributes for macro args.

### PR DESCRIPTION
**Pull request by ehuss**
This adds some validation for `path`, `crate_type`, and `recursion_limit` attributes so that they will now return an error if you attempt to pass a macro into them (such as `#[path = foo!()]`). Previously, the attribute would be completely ignored. These attributes are special because their values need to be known before/during expansion.

cc #87681

### CODE REVIEW

**Comment by ehuss** [comment]
I'm not too happy with this change, as it seems to add some specific validation that would probably better be part of something more general. However, attribute validation happens long after expansion, and I'm not aware of any pre-expansion validation where this could live.

**Comment by petrochenkov** [comment]
We actually validate some built-in attributes during expansion, those that are applied to macro calls and disappear during expansion.
```rust
macro_rules! m { () => () }

#[path] // error: malformed `path` attribute input
m!();

fn main() {}
```
This happens in `expand.rs / fn check_attributes`.

You can also reuse `validate_attr::check_meta` if you want to get the same wording as during regular attribute validation.

File: compiler/rustc_interface/src/util.rs [hunk 1]
```diff
@@ -528,6 +528,18 @@ pub(crate) fn check_attr_crate_type(
                         );
                     }
                 }
+            } else {
+                // This is here mainly to check for using a macro, such as
+                // #![crate_type = foo!()]. That is not supported since the
+                // crate type needs to be known very early in compilation long
+                // before expansion. Otherwise, validation would normally be
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/util.rs:535]: >crate type needs to be known very early in compilation long before expansion

Why tho?
This information seems irrelevant to parser or macro expander.

**Comment by petrochenkov** [comment]
This seems fine.
The point is to run validation before the attribute is used (that's why most uses simply ignore ill-formed attributes).
Since most attributes are used late, the validation is run soon after expansion, but for a few attributes that are used early the validation can be done immediately before use, don't see problem in that.

File: compiler/rustc_interface/src/util.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_interface/src/util.rs:535]: Ah, I see `StableCrateId::new` uses the crate types, although the use seems incorrect.

Otherwise it's only used before proc macro harness generation (to determine whether generate it or not), which happens only after all macros are expanded.

In other words, this is something that needs to be investigated.

Comment (petrochenkov) [reply compiler/rustc_interface/src/util.rs:535]: cc @bjorn3 
This is related to https://github.com/rust-lang/rust/pull/83744 (for which I don't see much motivation).

Comment (bjorn3) [reply compiler/rustc_interface/src/util.rs:535]: > Ah, I see StableCrateId::new uses the crate types, although the use seems incorrect.

This is necessary for backwards compatibility. Without it the following would give an error about a duplicate `StableCrateId`:

```bash
$ echo "" | rustc - --crate-type lib --crate-name foo
$ echo "extern crate foo; fn main() {}" | rustc - --crate-type bin --crate-name foo -L .
```

the error would be (forced it by using lib as crate type in both cases):

```
error[E0519]: the current crate is indistinguishable from one of its dependencies: it has the same crate-name `foo` and was compiled with the same `-C metadata` arguments. This will result in symbol conflicts between the two.
 --> <anon>:1:1
  |
1 | extern crate foo; fn main() {}
  | ^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

I would personally be happy if it were to be turned into an error as it complicates things, but it is a breaking change.

> Otherwise it's only used before proc macro harness generation (to determine whether generate it or not), which happens only after all macros are expanded.

Currently expansion has to happen before the `StableCrateId` can be set in the `Session` due to the dependency on expansion to determine all `#![crate_type = "..."]`. Once the future compat lint in #83744 becomes a hard error, it becomes possible to determine the final crate name and crate type for certain without any expansion by only parsing the crate level attributes, thus allowing the `StableCrateId` to be set immediately while creating the `Session`.

Comment (petrochenkov) [reply compiler/rustc_interface/src/util.rs:535]: >allowing the `StableCrateId` to be set immediately while creating the `Session`

Are there specific reasons to want `StableCrateId` be created together with the `Session`?

**Comment by ehuss** [comment]
I'm not sure if you meant that it should check in `InvocationCollector`, but I liked the idea so I ran with it.  WDYT?

I also fixed a minor issue with the suggested fix not checking if it was an inner attribute.

File: compiler/rustc_interface/src/util.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_interface/src/util.rs]: @bjorn3 
Could you copy this motivation into the description of https://github.com/rust-lang/rust/pull/83744?

**Comment by petrochenkov** [comment]
Nah, I only used `InvocationCollector` as an example, showing that we are already checking attributes early in some cases.

I liked the initial version of the PR better, TBH.
I'd just suggest to return to it, only use `validate_attr::check_meta` instead of hand-rolled checking.

**Comment by petrochenkov** [comment]
The inner attribute fix and moving the `emit_malformed_attribute` logic to a separate function are good though.

File: compiler/rustc_interface/src/util.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_interface/src/util.rs]: Done

**Comment by ehuss** [comment]
>  I'd just suggest to return to it, only use `validate_attr::check_meta` instead of hand-rolled checking.

Sure, no problem.  Unfortunately, that means adding a dependency from rustc_middle to rustc_parse, which does not seem good to me?

**Comment by bjorn3** [comment]
Can `validate_attr::check_meta` be moved to rustc_attr?

**Comment by petrochenkov** [comment]
Or `validate_attr::check_meta` can be called in `fn configure_and_expand` at call site of `get_recursion_limit`.

**Comment by ehuss** [comment]
> Or `validate_attr::check_meta` can be called in `fn configure_and_expand` at call site of `get_recursion_limit`.

Sure, I updated it to do the check in `configure_and_expand`.

**Comment by ehuss** [comment]
@bors retry

Error in `test_cat::prop_cat_rows` is xsv, which seems unlikely to be related here. A similar issue was noted a few years ago at https://github.com/rust-lang/rust/pull/65342#issuecomment-541349793

**Comment by alex-berger** [comment]
The following code used to compile up-to-and-including `nightly-2021-09-26`.

```rust
#[path = concat!(env!("OUT_DIR"), "/v1_22/mod.rs")]
mod v1_22;
```

And starting with `nightly-2021-09-27` it fails with:

```
error: malformed `path` attribute input
 --> ...
  |
2 | #[path = concat!(env!("OUT_DIR"), "/v1_22/mod.rs")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[path = "file"]`
```

Is this intended?

**Comment by ehuss** [comment]
The effect before was that the `#[path]` attribute was ignored.

Hm, so my expectation was that nobody would be using that, since the attribute is ignored and you would need a `v1_22.rs` file in the same directory to avoid a compilation error (and I would suspect most people would notice that it was using the wrong file).  Did you maybe have a `v1_22.rs` or `v1_22/mod.rs` file that was being used instead?

We could maybe downgrade it to a future-incompatibility warning if people feel this was too aggressive.

**Comment by alex-berger** [comment]
@ehuss indeed, there is a `$OUT_SIR/v1_22/model.rs` in place and I guess this is why it used to work. 

Personally, I am fine with this change, just wanted to understand why it used to work.
